File: trains-gdx-core/src/main/java/com/github/caay2000/trains/world/Train.java
package com.github.caay2000.trains.world;

import com.github.caay2000.trains.common.Position;

public class Train {

    private Route route;
    private Position position;
    private double speed;

    public Train(Route route, double speed) {
        this.route = route;
        this.position = new Position(route.getStart().getPosition());
        this.speed = speed;
    }

    public Route getRoute() {
        return route;
    }

    public Position getPosition() {
        return position;
    }

    public double getSpeed() {
        return speed;
    }

    public void move(float elapsed) {
        updatePosition(elapsed);
    }

    private void updatePosition(float elapsed) {
        float distance = position.distanceTo(route.getEnd().getPosition());
        if (speed * elapsed > distance) {
            calculateNextRoute();
            elapsed = calculateElpsedMissing(elapsed, distance);
        }
        position = position.move(route.getEnd().getPosition(), (float) (elapsed * speed));
    }

    private float calculateElpsedMissing(float elapsed, float distance) {
        distance = (float) (elapsed * speed) - distance;
        elapsed = (elapsed * distance) / (float) (speed * elapsed);
        return elapsed;
    }

    private void calculateNextRoute() {
        position = new Position(route.getEnd().getPosition());
        route = new Route(route.getEnd(), route.getStart());
    }

}

File: trains-gdx-core/src/main/kotlin/com/github/caay2000/trains/world/Train.kt
package com.github.caay2000.trains.world

class Train : Entity {

    override val position: Position
    private val speed: Float
    private var route: Route

    constructor(speed: Float, route: Route) {
        this.position = Position(route.start.position)
        this.speed = speed
        this.route = route
    }

    override fun update(delta: Float) {
        this.updatePosition(speed * delta)
    }

    private fun move(to: Location, distance: Float) {
    }

    private fun updatePosition(elapsed: Float) {
        var missing = elapsed
        val distance = position.distanceTo(route.end.position)
        if (speed * elapsed > distance) {
            calculateNextRoute()
            missing = calculateElapsedMissing(elapsed, distance)
        }
        position.move(route.end.position, missing * speed)
    }

    private fun calculateElapsedMissing(elapsed: Float, distance: Float): Float {
        val missing = (elapsed * speed) - distance
        return elapsed * missing / (speed * elapsed)
    }

    private fun calculateNextRoute() {
        position.translate(route.end.position)
        route = Route(route.end, route.start)
    }
}